sbol2/partshop.py

```diff
@@ -344,6 +344,55 @@ def attachFile(self, top_level_uri, filepath):
         msg = msg.format(response.status_code)
         raise SBOLError(SBOLErrorCode.SBOL_ERROR_BAD_HTTP_REQUEST, msg)
 
+    def downloadAttachment(self, attachment_uri, filepath='.'):
+        '''Download a file attachment from SynBioHub.
+        :param attachment_uri: The URI of the file to download
+        :param filepath: The local path or filename to which the file will be written
+        Returns None if successful.
+
+        Raises EnvironmentError if the filepath is invalid
+
+        Raises SBOLError with code SBOL_ERROR_HTTP_UNAUTHORIZED if it
+        there is an HTTP Unauthorized response.
+
+        Raises SBOLError with code SBOL_ERROR_BAD_HTTP_REQUEST on any
+        other HTTP error. The actual status code is embedded in the
+        string message.
+        '''
+        url = self._uri2url(attachment_uri)
+        url = posixpath.join(url, 'download')
+        filepath = os.path.expanduser(filepath)
+
+        # HTTP request headers
+        headers = {
+            'Accept': 'text/plain',
+            'X-authorization': self.key
+        }
+
+        # Issue GET request
+        response = requests.get(url, headers=headers)
+        if response.ok:
+            filename = response.headers['Content-Disposition']
+            filename = filename[22:-1]  # remove extraneous text: attachment; filename="
+            if os.path.isdir(filepath):
+                filepath = posixpath.join(filepath, filename)
+            with open(filepath, 'wb') as filehandle:
+                filehandle.write(response.content)
+            return
+        if response.status_code == http.HTTPStatus.UNAUTHORIZED:
+            # HTTP 401
+            msg = 'You must login with valid credentials before attaching a file'
```

Comment:
Copy paste error? Seems like the wrong error message.

Reply:
Changed to 'You must login with valid credentials before _downloading_ a file'